### PR TITLE
Support GIT lib for getting git commits per files

### PR DESCRIPTION
**Pull request by OsmondJiang**
support git lib for getting git commits per files

### CODE REVIEW

File: src/docfx/lib/git/models/GitHubUser.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs
```
Comment (yufeih) [inline src/docfx/lib/git/models/GitHubUser.cs]: remove this file

File: src/docfx/docfx.csproj [hunk 2]
```diff
@@ -5,15 +5,20 @@
...
 
   <ItemGroup>
     <AdditionalFiles Include="../../stylecop.json" />
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="LibGit2Sharp.NativeBinaries" Version="1.0.210" />
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.0" PrivateAssets="All" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Octokit" Version="0.29.0" />
```
Comment (yufeih) [inline src/docfx/docfx.csproj]: remove

File: src/docfx/lib/git/NativeMethods.cs [hunk 3]
```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Docs
```
Comment (yufeih) [inline src/docfx/lib/git/NativeMethods.cs]: Microsoft.Docs.Git ?

File: stylecop.ruleset [hunk 4]
```diff
@@ -42,10 +44,12 @@
     <Rule Id="CA1405" Action="Warning" />
     <Rule Id="CA1410" Action="Warning" />
     <Rule Id="CA1415" Action="Warning" />
+    <Rule Id="CA1806" Action="None" />
```
Comment (yufeih) [inline stylecop.ruleset]: xml comment for what CA1806 is? when Action="None"

File: src/docfx/lib/PathUtil.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide utils of path
+    /// </summary>
+    public static class PathUtil
```
Comment (yufeih) [inline src/docfx/lib/PathUtil.cs]: PathUtilTest ?

File: src/docfx/lib/git/models/GitHubUser.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/lib/git/models/GitHubUser.cs]: done

File: src/docfx/docfx.csproj (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/docfx.csproj]: done

File: stylecop.ruleset (hunk 4 above)
Comment (OsmondJiang) [reply stylecop.ruleset]: add comment

File: src/docfx/lib/PathUtil.cs (hunk 5 above)
Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: add

Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: add

File: src/docfx/docfx.csproj (hunk 2 above)
Comment (928PJY) [reply src/docfx/docfx.csproj]: 1. Can we add comments in below of the author but not add a new comment?
2. Change the Comment status but not add a new `done` comment?

---
In reply to: [179977756](https://github.com/dotnet/docfx/pull/2601#discussion_r179977756) [](ancestors = 179977756)

**Comment by 928PJY** [comment]
> }

nothing changed in this file?

---
Refers to: src/docfx/Program.cs:12 in bf9657d. [](commit_id = bf9657d250bf387e839a31a6a3396de55a790e21, deletion_comment = False)

File: src/docfx/lib/PathUtil.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+            return Normalize(path);
+        }
+
+        private static string Normalize(string path)
+        {
+            path = path.Replace('\\', '/');
+
+            var needReorder = false;
+            foreach (var c in path)
```
Comment (928PJY) [inline src/docfx/lib/PathUtil.cs]: >foreach (var c in path) [](start = 12, length = 23)

string.IndexOf()?

File: src/docfx/lib/PathUtil.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            }
+            if (!needReorder)
+            {
+                return path;
+            }
+
+            var parentCount = 0;
+            var rooted = path[0] == '/';
+            var stack = new List<string>();
+            foreach (var segment in path.Split('/', '\\'))
```
Comment (928PJY) [inline src/docfx/lib/PathUtil.cs]: >'\\' [](start = 52, length = 4)

remove?

Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: why?

---
In reply to: [179994083](https://github.com/dotnet/docfx/pull/2601#discussion_r179994083) [](ancestors = 179994083)

File: src/docfx/lib/PathUtil.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+                }
+            }
+            if (!needReorder)
+            {
+                return path;
+            }
+
+            var parentCount = 0;
+            var rooted = path[0] == '/';
+            var stack = new List<string>();
```
Comment (928PJY) [inline src/docfx/lib/PathUtil.cs]: >stack [](start = 16, length = 5)

why  do we call it `stack`?

Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: you can check below operations which make it like a stack

---
In reply to: [179995131](https://github.com/dotnet/docfx/pull/2601#discussion_r179995131) [](ancestors = 179995131)

File: src/docfx/lib/git/GitUtil.cs [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+                {
+                    return repo;
+                }
+                repo = Path.GetDirectoryName(repo);
+            }
+            return repo;
+        }
+
+        /// <summary>
+        /// Get git repo information, will change to git exe later
```
Comment (fenxu) [inline src/docfx/lib/git/GitUtil.cs]: >will change to git exe later [](start = 38, length = 28)

Why don't we do it right now? I don't think we need to rush.

Comment (OsmondJiang) [reply src/docfx/lib/git/GitUtil.cs]: we don't need it right now, i will remove it from this pr

---
In reply to: [179996870](https://github.com/dotnet/docfx/pull/2601#discussion_r179996870) [](ancestors = 179996870)

File: src/docfx/lib/PathUtil.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        /// <summary>
+        /// Create a relative path from one path to another file.
+        /// Use this over <see cref="Path.GetRelativePath(string, string)"/> when
+        /// <paramref name="fileRelativeTo"/> is a file.
+        /// </summary>
+        /// <param name="fileRelativeTo">The file path relative to</param>
+        /// <param name="path">The original path </param>
+        /// <returns>The relative path</returns>
+        public static string GetRelativePathToFile(string fileRelativeTo, string path)
```
Comment (fenxu) [inline src/docfx/lib/PathUtil.cs]: >string fileRelativeTo, string path [](start = 51, length = 34)

Also need to check invalid for those two

File: src/docfx/lib/PathUtil.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+            return Path.GetRelativePath(directory, path);
+        }
+
+        /// <summary>
+        /// A normalized folder always ends with `/`, does not contain `\` and does not have consegative `.` or `/`.
+        /// </summary>
+        /// <param name="path">The folder path want to be normalized</param>
+        /// <returns>The normalized folder path</returns>
+        public static string NormalizeFolder(string path)
```
Comment (fenxu) [inline src/docfx/lib/PathUtil.cs]: >path [](start = 52, length = 4)

Check for invalid file name

File: src/docfx/lib/PathUtil.cs [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+        /// <summary>
+        /// A normalized file cannot end with `/`, does not contain `\` and does not have consegative `.` or `/`.
+        /// </summary>
+        /// <param name="path">The file path want to be normalized</param>
+        /// <returns>The normalized file path</returns>
+        public static string NormalizeFile(string path)
+        {
+            Debug.Assert(Path.GetDirectoryName(path).IndexOfAny(Path.GetInvalidPathChars()) < 0);
+            Debug.Assert(Path.GetFileName(path).IndexOfAny(Path.GetInvalidFileNameChars()) < 0);
```
Comment (fenxu) [inline src/docfx/lib/PathUtil.cs]: Extract this to a common method that check invalid path? It's common logic.

File: src/docfx/lib/git/GitUtil.cs [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+                result[i] = GetCommitsByPath(files[i], trees, pathToParentByRef, commits);
+                progress?.Invoke(Interlocked.Increment(ref done), total);
+            });
+            NativeMethods.GitRepositoryFree(repo);
+            return result;
+        }
+
+        private static FileMap BuildPathToParentPath(List<string> files)
+        {
+            var res = new HashSet<string>(files.Count + 128);
```
Comment (fenxu) [inline src/docfx/lib/git/GitUtil.cs]: >128 [](start = 56, length = 3)

Why use 128?

File: src/docfx/lib/PathUtil.cs [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+            if (string.IsNullOrEmpty(directory))
+            {
+                return path;
+            }
+
+            return Path.GetRelativePath(directory, path);
+        }
+
+        /// <summary>
+        /// A normalized folder always ends with `/`, does not contain `\` and does not have consegative `.` or `/`.
```
Comment (superyyrrzz) [inline src/docfx/lib/PathUtil.cs]: >consegative [](start = 93, length = 11)

consecutive?

File: src/docfx/lib/PathUtil.cs [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            var needReorder = false;
+            if (path.IndexOf('.') != -1)
+            {
+                needReorder = true;
+            }
+
+            if (!needReorder)
+            {
+                return path;
+            }
```
Comment (superyyrrzz) [inline src/docfx/lib/PathUtil.cs]: directly `if (path.IndexOf('.') == -1) return path;` ?

File: src/docfx/lib/PathUtil.cs [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        /// <summary>
+        /// A normalized file cannot end with `/`, does not contain `\` and does not have consegative `.` or `/`.
+        /// </summary>
+        /// <param name="path">The file path want to be normalized</param>
+        /// <returns>The normalized file path</returns>
+        public static string NormalizeFile(string path)
+        {
+            Debug.Assert(Path.GetDirectoryName(path).IndexOfAny(Path.GetInvalidPathChars()) < 0);
```
Comment (fenxu) [inline src/docfx/lib/PathUtil.cs]: >.IndexOfAny(Path.GetInvalidPathChars()) < 0 [](start = 52, length = 43)

Remove this

File: src/docfx/lib/PathUtil.cs [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            else
+            {
+                while (parentCount-- > 0)
+                {
+                    res.Append("../");
+                }
+            }
+
+            var i = 0;
```
Comment (superyyrrzz) [inline src/docfx/lib/PathUtil.cs]: >i [](start = 16, length = 1)

use bool? seems only care about whether it's 0.

File: src/docfx/lib/PathUtil.cs [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+                needReorder = true;
+            }
+
+            if (!needReorder)
+            {
+                return path;
+            }
+
+            var parentCount = 0;
+            var rooted = path[0] == '/';
```
Comment (superyyrrzz) [inline src/docfx/lib/PathUtil.cs]: >var rooted = path[0] == '/'; [](start = 12, length = 28)

what about assert the path is not rooted? Now `/../a` will be normalized to `/a`

File: src/docfx/lib/PathUtil.cs [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide utils of path
+    /// </summary>
+    public static class PathUtil
```
Comment (vicancy) [inline src/docfx/lib/PathUtil.cs]: >PathUtil [](start = 24, length = 8)

naming: use full name: PathUtility

File: src/docfx/lib/PathUtil.cs [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            Debug.Assert(folder.Length > 0 && folder[folder.Length - 1] == '/');
+
+            return folder == "./" || path.StartsWith(folder, StringComparison.OrdinalIgnoreCase);
+        }
+
+        /// <summary>
+        /// Create a relative path from one path to another file.
+        /// Use this over <see cref="Path.GetRelativePath(string, string)"/> when
+        /// <paramref name="fileRelativeTo"/> is a file.
```
Comment (928PJY) [inline src/docfx/lib/PathUtil.cs]: fix this

File: src/docfx/lib/git/GitUtil.cs [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+        /// <param name="progress">The processing progress</param>
+        /// <returns>A collection of git commits</returns>
+        public static unsafe List<GitCommit>[] GetCommits(string repoPath, List<string> files, Action<int, int> progress = null)
+        {
+            if (string.IsNullOrEmpty(repoPath))
+            {
+                throw new ArgumentNullException(nameof(repoPath));
+            }
+
+            Debug.Assert(files.All(file => file == PathUtil.NormalizeFile(file)));
```
Comment (vicancy) [inline src/docfx/lib/git/GitUtil.cs]: >Debug.Assert(files.All(file => file == PathUtil.NormalizeFile(file))); [](start = 12, length = 70)

If `NormalizeFile` is used only in Debug.Assert, shall we exclude PathUtil in this PR and focus on GitUtility and here we check if it contains unwanted characters such as `..` or `\`?

File: src/docfx/lib/git/models/GitCommit.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+{
+    /// <summary>
+    /// The git commit information
+    /// </summary>
+    public class GitCommit
+    {
+        /// <summary>
+        /// Gets or sets the git commmit author name
+        /// </summary>
+        public string AuthorName { get; set; }
```
Comment (vicancy) [inline src/docfx/lib/git/models/GitCommit.cs]: >AuthorName [](start = 22, length = 10)

Should it contain both `Committer` and `Author`?

File: src/docfx/lib/git/NativeMethods.cs [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Docs.Git
+{
+    internal static class NativeMethods
+    {
+        private const string LibName = "git2-6311e88";
```
Comment (vicancy) [inline src/docfx/lib/git/NativeMethods.cs]: >LibName [](start = 29, length = 7)

does it mean that we should embed the dll in our solution? Am not seeing that..

File: src/docfx/lib/git/GitUtil.cs [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+    /// Provide git operations
+    /// </summary>
+    public static class GitUtil
+    {
+        /// <summary>
+        /// Find git repo directory
+        /// </summary>
+        /// <param name="path">The git repo entry point</param>
+        /// <returns>The git repo root path</returns>
+        public static string FindRepo(string path)
```
Comment (superyyrrzz) [inline src/docfx/lib/git/GitUtil.cs]: >path [](start = 45, length = 4)

check invalid chars?

File: test/docfx.Test/lib/GitUtilTest.cs [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+        [InlineData("README.md")]
+        public static void GetCommitsSameAsGitLog(string file)
+        {
+            var repo = GitUtil.FindRepo(Path.GetFullPath(file));
+            var pathToRepo = PathUtil.NormalizeFile(file);
+            var exe = GetContributorsGitExe(repo, pathToRepo).ToList();
+            var lib = GitUtil.GetCommits(repo, new List<string> { pathToRepo })[0].ToList();
+            Assert.Equal(JsonConvert.SerializeObject(exe), JsonConvert.SerializeObject(lib));
+
+            GitCommit[] GetContributorsGitExe(string cwd, string path)
```
Comment (vicancy) [inline test/docfx.Test/lib/GitUtilTest.cs]: >GetContributorsGitExe [](start = 24, length = 21)

Move out

File: src/docfx/lib/PathUtil.cs (hunk 17 above)
Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: i believe both would be ok

---
In reply to: [180007222](https://github.com/dotnet/docfx/pull/2601#discussion_r180007222) [](ancestors = 180007222)

File: src/docfx/lib/PathUtil.cs (hunk 20 above)
Comment (OsmondJiang) [reply src/docfx/lib/PathUtil.cs]: it's a ref in the summary

---
In reply to: [180008851](https://github.com/dotnet/docfx/pull/2601#discussion_r180008851) [](ancestors = 180008851)

File: src/docfx/lib/git/models/GitCommit.cs (hunk 22 above)
Comment (OsmondJiang) [reply src/docfx/lib/git/models/GitCommit.cs]: what will be used finally?

---
In reply to: [180009486](https://github.com/dotnet/docfx/pull/2601#discussion_r180009486) [](ancestors = 180009486)

File: src/docfx/lib/git/GitUtil.cs (hunk 21 above)
Comment (OsmondJiang) [reply src/docfx/lib/git/GitUtil.cs]: it not only used here, will add has invalid chars method later

---
In reply to: [180009088](https://github.com/dotnet/docfx/pull/2601#discussion_r180009088) [](ancestors = 180009088)

File: test/docfx.Test/lib/GitUtilTest.cs [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.Docs.Test
+{
+    public static class GitUtilTest
```
Comment (vicancy) [inline test/docfx.Test/lib/GitUtilTest.cs]: >GitUtil [](start = 24, length = 7)

GitUtility

**Pull request merged**